sycl/include/syclcompat/util.hpp

```diff
@@ -199,6 +199,38 @@ inline unsigned int byte_level_permute(unsigned int a, unsigned int b,
   return ret;
 }
 
+/// \brief The functions performs bitwise logical operations on three input
+/// values of \p a, \p b and \p c based on the specified 8-bit truth table \p
+/// lut and return the result
+///
+/// \param [in] a Input value
+/// \param [in] b Input value
+/// \param [in] c Input value
+/// \param [in] lut truth table for looking up
+/// \returns The result
+inline uint32_t lop3(uint32_t a, uint32_t b, uint32_t c, uint8_t lut) {
+  uint32_t result = 0;
+
```

Comment:
It is better to use the optimized instructions for backends when available, so that translation does not reduce performance wrt cuda.

```suggestion

#if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__)
asm volatile("lop3.b32 %0, %1, %2, %3, %4;"
             : "=r"(result)
             : "r"(a), "r"(b), "r"(c), "r"(lut));
#else
```

See later corresponding #endif suggestion

Reply:
> It is better to use the optimized instructions for backends when available, so that translation does not reduce performance wrt cuda.
> 
> See later corresponding #endif suggestion

I have refined the helper function. As to keeping asm PTX code in the helper function for SYCL CUDA backend, I think it is not necessary, as SYCLomatic has provided the option “--optimize-migration.”  If this option is specified during migration, the PTX asm instruction will be kept in the migrated code, here is the demo case https://github.com/oneapi-src/SYCLomatic/blob/821800fb720a82403a4488d90ea8233cca45b918/clang/test/dpct/asm/optimize.cu#L11

Reply:
Doesn't this imply that 
- users should be using syclomatic conversion potentially twice in cases where they want to support both cuda supported (optimized) and e.g. l0 or other backends? Then they have two non-portable codes to maintain: Doesn't this go completely against the philosophy of oneapi
- Users won't ever write sycl::compat code themselves, since the above point implies that they won't know what to write for their particular target unless they use the automatic translation?

The above situation also implies that sycl::compat will be able to generally maintain performance (non portable) for at most two backends (cuda and l0), since it dissuades further backend specific optimized implementation (in a portable) for any other backend e.g. HIP.

I don't understand how any of this makes sense.

Reply:
It is worth clarifying that of course sometimes it won't make sense to write the asm! The example you gave:
```
// CHECK-NEXT: #if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__)
// CHECK-NEXT:   asm("mov.s32 %0, %1;" : "=r"(a) : "r"(b));
// CHECK-NEXT: #else
// CHECK-NEXT:   a = b;
// CHECK-NEXT: #endif
```

Is actually never going to be better than just `a = b` since the compiler takes care of the lowering to the ptx instruction in this case.

There is actually a third case e.g. `mov.b32` which is a ptx instruction that in some very specialized cases might be good to write directly (since an appropriate lowering might not be available in the compiler).

But these are different cases to the one in this PR: which is a very specialized/optimized ptx instruction that does not have a compiler lowering (and probably it doesn't make sense to add one).

But in all such cases it is appropriate to deal with these on a case by case basis within sycl::compat (or other sycl headers).

Reply:
> asm volatile("lop3.b32 %0, %1, %2, %3, %4;"
>              : "=r"(result)
>              : "r"(a), "r"(b), "r"(c), "r"(lut));

Okay,  accept your advice, use ASM PTX instruct for SYCL CUDA backend.

Reply:
> Okay, accept your advice, use ASM PTX instruct for SYCL CUDA backend.

Okay, accept your advice, use ASM PTX instructions for SYCL CUDA backend,  I have addressed your comments in the updated commit, pls take a look.

Reply:
> Doesn't this imply that
> 
> * users may use syclomatic conversion potentially twice in cases where they want to support both cuda supported (optimized) and e.g. l0 or other backends (but without complex preprocessor directive directly in source code)? Then they have either two non-portable codes to maintain, or a code with lots of #ifdefs in for the cuda path: Doesn't this go completely against the philosophy of oneapi?
> * Users (that want cuda performance) won't ever write sycl::compat code themselves (i.e. via reading documentation for functions they need), since the above point implies that they won't know what to write for their particular target unless they use the automatic translation?
> 
> The above situation also implies that sycl::compat will be able to generally maintain performance (non portably) for at most two backends (cuda and l0), since it dissuades further backend specific optimized implementation (in a portable manner) for any other backend e.g. HIP.
> 
> Isn't it better to provide the preprocessor #if #else abstraction inside of the sycl (compat) functions to enable simpler portable code?

Okay, accept your advice and use ASM PTX instructions for the SYCL CUDA backend; I have addressed your comments in the updated commit; pls take a look.

Reply:
> Doesn't this imply that
> 
> * users may use syclomatic conversion potentially twice in cases where they want to support both cuda supported (optimized) and e.g. l0 or other backends (but without complex preprocessor directive directly in source code)? Then they have either two non-portable codes to maintain, or a code with lots of #ifdefs in for the cuda path: Doesn't this go completely against the philosophy of oneapi?
> * Users (that want cuda performance) won't ever write sycl::compat code themselves (i.e. via reading documentation for functions they need), since the above point implies that they won't know what to write for their particular target unless they use the automatic translation?
> 
> The above situation also implies that sycl::compat will be able to generally maintain performance (non portably) for at most two backends (cuda and l0), since it dissuades further backend specific optimized implementation (in a portable manner) for any other backend e.g. HIP.
> 
> Isn't it better to provide the preprocessor #if #else abstraction inside of the sycl (compat) functions to enable simpler portable code?

Okay, accept your advice and use ASM PTX instructions for the SYCL CUDA backend; I have addressed your comments in the updated commit; pls take a look.